docs/Training-Generalization-Learning.md

```diff
@@ -0,0 +1,123 @@
+# Training Generalized Reinforcement Learning agents
+
+Reinforcement learning has a rather unique setup as opposed to supervised and
+unsupervised learning. Agents here are trained and tested on the same exact 
+environment, which is analogous to a model being trained and tested on an 
+identical dataset in supervised learning! This setting results in overfitting; 
+the inability of the agent to generalize to slight tweaks or variations in the 
+environment. This is problematic in instances when environments are randomly 
+instantiated with varying properties. To make agents more robust, we train an 
+agent over multiple variations of the environment. The agent is trained with 
+the intent that it learns to maintain a minimum performance regardless of the 
+environment variant and that it generalizes to maintain this in unseen future 
+variants of the environment.
+
+Ball scale of 0.5          |  Ball scale of 4
+:-------------------------:|:-------------------------:
+![](images/3dball_small.png)  |  ![](images/3dball_big.png)
+
+_Variations of the 3D Ball environment._
+
+To vary environments, we first decide what parameters to vary in an
+environment. These parameters are known as `Reset Parameters`. In the 3D ball 
+environment example displayed in the figure above, the reset parameters are `gravity`, `ball_mass` and `ball_scale`.
+
+
+## How-to
+
+For generalization training, we need to provide a way to modify the environment 
+by supplying a set of reset parameters. This provision can be either 
```

Comment:
To make this clearer, I'd break this into two paragraphs. First: "For generalization training, we need to provide a way to ... reset parameters, and vary them over time. The parameters could be chosen either deterministically or randomly."

2nd paragraph: "This is done by assigning each reset parameter a **sampler**, which (insert description of what a sampler does). ..."